test/T42.stl.names.multi.ref

```diff
@@ -22,34 +22,26 @@
 
 void bind_std_stl_function(std::function< pybind11::module &(std::string const &namespace_) > &M)
 {
-	{ // std::binary_function file:bits/stl_function.h line:114
+	{ // std::binary_function file:bits/stl_function.h line:118
 		pybind11::class_<std::binary_function<float,float,bool>, std::shared_ptr<std::binary_function<float,float,bool>>> cl(M("std"), "binary_function_float_float_bool_t", "");
 		cl.def( pybind11::init( [](std::binary_function<float,float,bool> const &o){ return new std::binary_function<float,float,bool>(o); } ) );
 		cl.def( pybind11::init( [](){ return new std::binary_function<float,float,bool>(); } ) );
-		cl.def("assign", (struct std::binary_function<float, float, bool> & (std::binary_function<float,float,bool>::*)(const struct std::binary_function<float, float, bool> &)) &std::binary_function<float, float, bool>::operator=, "C++: std::binary_function<float, float, bool>::operator=(const struct std::binary_function<float, float, bool> &) --> struct std::binary_function<float, float, bool> &", pybind11::return_value_policy::automatic, pybind11::arg(""));
 	}
-	{ // std::equal_to file:bits/stl_function.h line:204
+	{ // std::equal_to file:bits/stl_function.h line:351
 		pybind11::class_<std::equal_to<float>, std::shared_ptr<std::equal_to<float>>, std::binary_function<float,float,bool>> cl(M("std"), "equal_to_float_t", "");
 		cl.def( pybind11::init( [](){ return new std::equal_to<float>(); } ) );
 		cl.def("__call__", (bool (std::equal_to<float>::*)(const float &, const float &) const) &std::equal_to<float>::operator(), "C++: std::equal_to<float>::operator()(const float &, const float &) const --> bool", pybind11::arg("__x"), pybind11::arg("__y"));
-		cl.def("assign", (struct std::binary_function<float, float, bool> & (std::binary_function<float,float,bool>::*)(const struct std::binary_function<float, float, bool> &)) &std::binary_function<float, float, bool>::operator=, "C++: std::binary_function<float, float, bool>::operator=(const struct std::binary_function<float, float, bool> &) --> struct std::binary_function<float, float, bool> &", pybind11::return_value_policy::automatic, pybind11::arg(""));
```

Comment:
I think it would be good if we decide to always produce reference results on CentOS-7 with LLVM-6 because otherwise we will have meaningless stream of changes when  `assign` and constructor operators got deleted and inserted. 

Also, producing reference results on non-reference platform while not triggering `diffbinder` still do break `self-test.py` tests. So could you please regenerate results results on CentOS-7 with LLVM-6?

Thanks,

Reply:
Hi @lyskov ,
First,
> otherwise we will have meaningless stream of changes
>Also, producing reference results on non-reference platform ...
completely agree.


> regenerate results results on CentOS-7 with LLVM-6?
Done. And there is still a difference with the references.
Also, one can see that the current reference in master is problematic just from this include:

```
#include <bits/std_function.h> // std::_Manager_operation
```
it really should be 
```
#include <functional> // std::_Manager_operation
```

The ``bits/std_function.h`` include does not even exist on CentOS7 ( gcc-4.8.5) with clang/llvm-6.0. I mean at all.
 It comes from the gcc in devtoolset-7+. Most likely that happens because when you generate these references using devtoolset and
binder picks up the stl from there. 
(+I'm not sure w/o searching for docs, but everything in the  ``bits/`` directory is implementation-dependent, I think. )

Would that make sense to add the path of stl headers to the python scripts explicitely?


Best regards,

Andrii

Reply:
> Would that make sense to add the path of stl headers to the python scripts explicitely?

-- this would not help. The issue is as you pointed out that some STL versions/implementations do have multiple layers of includes so "final" include for type/function/class will not necessary will be one that specified by C++ standard. We are addressing this in `add_relevant_include_for_decl` function (see 
https://github.com/RosettaCommons/binder/blob/master/source/type.cpp#L98) by mapping standard types to known C++11 includes and by mapping known non-standard include path into standard one. This approach however require adjustment from time to time as new failure cases arise.

Reply:
OK. I saw it, but did not put enough attention to understand why that map is there.
Would that make it possible to use another, more automatic approach?

Like if there is a tree of includes, some includes should be made special == the includes from standard.  And for every funny internal type the appropriate #include  would be just "navigate in the include tree till the first ancestor that is stl  include".
Of course, there should be a list of stl includes, but that is kind of simple as it is in standard.

Best regards,
Andrii

Reply:
Yes, i was thinking about this too, but so far it remains unclear how to extract include-chain info from LLVM AST. If you know of the examples of how to do it please let me know. Include mapping problem still remain as one of the unsolved issues so i am eager to revisit this when info of how to extract include chains became available. Thanks,

Reply:
I don't know yet. The LLVM  does not look that nice if to try to actually do  something.